File: backend/src/models/crm.py
from pydantic import BaseModel, Field, field_validator, model_validator
from typing import Optional, List, Any
from datetime import datetime
from uuid import UUID

from .labeled_contact_items import LabeledEmailItem, LabeledPhoneItem

class CustomerAttachmentItem(BaseModel):
    url: str
    original_filename: Optional[str] = None
    s3_key: Optional[str] = None

# Customer Models
class CustomerBase(BaseModel):
    firstName: str = Field(..., min_length=1, max_length=100)
    lastName: str = Field(..., min_length=1, max_length=100)
    email: Optional[str] = Field(None, pattern=r"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$")
    phone: Optional[str] = Field(None, max_length=20)

    @field_validator("email", mode="before")
    @classmethod
    def empty_email_to_none(cls, v):
        if v == "" or v is None:
            return None
        return v
    mobile: Optional[str] = Field(None, max_length=20)
    cnic: Optional[str] = Field(None, max_length=15)
    dateOfBirth: Optional[datetime] = None
    gender: Optional[str] = Field(None, pattern=r"^(male|female|other)$")
    address: Optional[str] = None
    city: Optional[str] = Field(None, max_length=100)
    state: Optional[str] = Field(None, max_length=100)
    country: Optional[str] = Field(default="Pakistan", max_length=100)
    postalCode: Optional[str] = Field(None, max_length=20)
    customerType: Optional[str] = Field(default="individual", pattern=r"^(individual|business)$")
    customerStatus: Optional[str] = Field(default="active", pattern=r"^(active|inactive|blocked)$")
    creditLimit: Optional[float] = Field(default=0.0, ge=0)
    currentBalance: Optional[float] = Field(default=0.0)
    paymentTerms: Optional[str] = Field(default="Cash", pattern=r"^(Credit|Card|Cash|Due Payments|immediate|net30|net60)$")
    assignedToId: Optional[UUID] = None
    notes: Optional[str] = None
    description: Optional[str] = None
    tags: Optional[List[str]] = Field(default_factory=list)
    attachments: List[CustomerAttachmentItem] = Field(default_factory=list)
    image_url: Optional[str] = None
    emails: List[LabeledEmailItem] = Field(default_factory=list)
    phones: List[LabeledPhoneItem] = Field(default_factory=list)

    @field_validator("attachments", mode="before")
    @classmethod
    def normalize_attachments(cls, v):
        if v is None:
            return []
        out = []
        for item in v:
            if isinstance(item, str):
                out.append({"url": item})
            elif isinstance(item, dict):
                out.append(item)
        return out

    @field_validator("emails", mode="before")
    @classmethod
    def normalize_emails(cls, v):
        if v is None:
            return []
        out = []
        for item in v:
            if isinstance(item, dict):
                out.append(item)
        return out

    @field_validator("phones", mode="before")
    @classmethod
    def normalize_phones(cls, v):
        if v is None:
            return []
        out = []
        for item in v:
            if isinstance(item, dict):
                out.append(item)
        return out

class CustomerCreate(CustomerBase):
    pass

class CustomerUpdate(BaseModel):
    firstName: Optional[str] = Field(None, min_length=1, max_length=100)
    lastName: Optional[str] = Field(None, min_length=1, max_length=100)
    email: Optional[str] = Field(None, pattern=r"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$")
    phone: Optional[str] = Field(None, max_length=20)
    mobile: Optional[str] = Field(None, max_length=20)
    cnic: Optional[str] = Field(None, max_length=15)
    dateOfBirth: Optional[datetime] = None
    gender: Optional[str] = Field(None, pattern=r"^(male|female|other)$")
    address: Optional[str] = None
    city: Optional[str] = Field(None, max_length=100)
    state: Optional[str] = Field(None, max_length=100)
    country: Optional[str] = Field(None, max_length=100)
    postalCode: Optional[str] = Field(None, max_length=20)
    customerType: Optional[str] = Field(None, pattern=r"^(individual|business)$")
    customerStatus: Optional[str] = Field(None, pattern=r"^(active|inactive|blocked)$")
    creditLimit: Optional[float] = Field(None, ge=0)
    currentBalance: Optional[float] = None
    paymentTerms: Optional[str] = Field(None, pattern=r"^(Credit|Card|Cash|Due Payments|immediate|net30|net60)$")
    assignedToId: Optional[UUID] = None
    notes: Optional[str] = None
    description: Optional[str] = None
    tags: Optional[List[str]] = None
    attachments: Optional[List[CustomerAttachmentItem]] = None
    image_url: Optional[str] = None
    emails: Optional[List[LabeledEmailItem]] = None
    phones: Optional[List[LabeledPhoneItem]] = None

class CustomerResponse(CustomerBase):
    id: UUID
    customerId: str
    tenant_id: UUID
    isActive: bool
    createdAt: datetime
    updatedAt: datetime
    
    @model_validator(mode='before')
    @classmethod
    def hydrate_customer_orm(cls, data: Any):
        if data is None or not hasattr(data, '_sa_instance_state'):
            return data
        from sqlalchemy.inspection import inspect as sa_inspect
        c = data
        emails = list(c.emails or [])
        if not emails and getattr(c, 'email', None):
            em = (c.email or '').strip()
            if em:
                emails = [{"value": em, "label": "personal"}]
        phones = list(c.phones or [])
        if not phones:
            if getattr(c, 'phone', None) and str(c.phone).strip():
                phones.append({"value": str(c.phone).strip(), "label": "work"})
            if getattr(c, 'mobile', None) and str(c.mobile).strip():
                phones.append({"value": str(c.mobile).strip(), "label": "personal"})
        out = {}
        for attr in sa_inspect(c).mapper.column_attrs:
            out[attr.key] = getattr(c, attr.key)
        out["emails"] = emails
        out["phones"] = phones
        return out

    class Config:
        from_attributes = True

class CustomersListResponse(BaseModel):
    customers: List[CustomerResponse]
    total: int

class CustomerStatsResponse(BaseModel):
    total_customers: int
    active_customers: int
    inactive_customers: int
    blocked_customers: int
    individual_customers: int
    business_customers: int
    recent_customers: int


class GuarantorBase(BaseModel):
    name: str
    mobile: Optional[str] = None
    cnic: Optional[str] = None
    residential_address: Optional[str] = None
    official_address: Optional[str] = None
    occupation: Optional[str] = None
    relation: Optional[str] = None
    display_order: Optional[int] = 0


class GuarantorCreate(GuarantorBase):
    pass


class GuarantorUpdate(BaseModel):
    name: Optional[str] = None
    mobile: Optional[str] = None
    cnic: Optional[str] = None
    residential_address: Optional[str] = None
    official_address: Optional[str] = None
    occupation: Optional[str] = None
    relation: Optional[str] = None
    display_order: Optional[int] = None


class GuarantorResponse(GuarantorBase):
    id: UUID
    tenant_id: UUID
    customer_id: UUID
    createdAt: datetime
    updatedAt: datetime
    
    class Config:
        from_attributes = True

File: mobile/src/models/crm.ts
export enum LeadStatus {
  NEW = 'new',
  CONTACTED = 'contacted',
  QUALIFIED = 'qualified',
  PROPOSAL_SENT = 'proposal_sent',
  NEGOTIATION = 'negotiation',
  WON = 'won',
  LOST = 'lost',
}

export enum LeadSource {
  WEBSITE = 'website',
  REFERRAL = 'referral',
  SOCIAL_MEDIA = 'social_media',
  EMAIL_CAMPAIGN = 'email_campaign',
  COLD_CALL = 'cold_call',
  TRADE_SHOW = 'trade_show',
  PARTNER = 'partner',
  OTHER = 'other',
}

export enum OpportunityStage {
  PROSPECTING = 'prospecting',
  QUALIFICATION = 'qualification',
  PROPOSAL = 'proposal',
  NEGOTIATION = 'negotiation',
  CLOSED_WON = 'closed_won',
  CLOSED_LOST = 'closed_lost',
}

export enum ContactType {
  LEAD = 'lead',
  CUSTOMER = 'customer',
  PARTNER = 'partner',
  VENDOR = 'vendor',
  OTHER = 'other',
}

export enum ActivityType {
  CALL = 'call',
  EMAIL = 'email',
  MEETING = 'meeting',
  TASK = 'task',
  NOTE = 'note',
  PROPOSAL = 'proposal',
  CONTRACT = 'contract',
}

export enum CompanySize {
  STARTUP = 'startup',
  SMALL = 'small',
  MEDIUM = 'medium',
  LARGE = 'large',
  ENTERPRISE = 'enterprise',
}

export enum Industry {
  TECHNOLOGY = 'technology',
  HEALTHCARE = 'healthcare',
  FINANCE = 'finance',
  MANUFACTURING = 'manufacturing',
  RETAIL = 'retail',
  EDUCATION = 'education',
  REAL_ESTATE = 'real_estate',
  CONSULTING = 'consulting',
  OTHER = 'other',
}

export interface Lead {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  phone?: string;
  company?: string;
  jobTitle?: string;
  status: LeadStatus;
  source: LeadSource;
  assignedTo?: string;
  notes?: string;
  tags: string[];
  score: number;
  budget?: number;
  timeline?: string;
  convertedToContact?: string;
  convertedToOpportunity?: string;
  lastContactDate?: string;
  nextFollowUpDate?: string;
  tenantId: string;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
}

export interface LeadCreate {
  firstName: string;
  lastName: string;
  email: string;
  phone?: string;
  company?: string;
  jobTitle?: string;
  status?: LeadStatus;
  source?: LeadSource;
  assignedTo?: string;
  notes?: string;
  tags?: string[];
  score?: number;
  budget?: number;
  timeline?: string;
}

export interface LeadUpdate {
  firstName?: string;
  lastName?: string;
  email?: string;
  phone?: string;
  company?: string;
  jobTitle?: string;
  status?: LeadStatus;
  source?: LeadSource;
  assignedTo?: string;
  notes?: string;
  tags?: string[];
  score?: number;
  budget?: number;
  timeline?: string;
}

export interface Contact {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  phone?: string;
  mobile?: string;
  jobTitle?: string;
  department?: string;
  companyId?: string;
  type: ContactType;
  notes?: string;
  tags: string[];
  isActive: boolean;
  lastContactDate?: string;
  nextFollowUpDate?: string;
  tenantId: string;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
}

export interface ContactCreate {
  firstName: string;
  lastName: string;
  email: string;
  phone?: string;
  mobile?: string;
  jobTitle?: string;
  department?: string;
  companyId?: string;
  type?: ContactType;
  notes?: string;
  tags?: string[];
  isActive?: boolean;
}

export interface ContactUpdate {
  firstName?: string;
  lastName?: string;
  email?: string;
  phone?: string;
  mobile?: string;
  jobTitle?: string;
  department?: string;
  companyId?: string;
  type?: ContactType;
  notes?: string;
  tags?: string[];
  isActive?: boolean;
}

export interface Company {
  id: string;
  name: string;
  industry?: Industry;
  size?: CompanySize;
  website?: string;
  phone?: string;
  email?: string;
  address?: string;
  city?: string;
  state?: string;
  country?: string;
  postalCode?: string;
  description?: string;
  notes?: string;
  tags: string[];
  isActive: boolean;
  annualRevenue?: number;
  employeeCount?: number;
  foundedYear?: number;
  tenantId: string;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
}

export interface CompanyCreate {
  name: string;
  industry?: Industry;
  size?: CompanySize;
  website?: string;
  phone?: string;
  address?: string;
  city?: string;
  state?: string;
  country?: string;
  postalCode?: string;
  description?: string;
  notes?: string;
  tags?: string[];
  isActive?: boolean;
  annualRevenue?: number;
  employeeCount?: number;
  foundedYear?: number;
}

export interface CompanyUpdate {
  name?: string;
  industry?: Industry;
  size?: CompanySize;
  website?: string;
  phone?: string;
  address?: string;
  city?: string;
  state?: string;
  country?: string;
  postalCode?: string;
  description?: string;
  notes?: string;
  tags?: string[];
  isActive?: boolean;
  annualRevenue?: number;
  employeeCount?: number;
  foundedYear?: number;
}

export interface Opportunity {
  id: string;
  title: string;
  description?: string;
  stage: OpportunityStage;
  amount?: number;
  probability: number;
  expectedCloseDate?: string;
  leadId?: string;
  contactId?: string;
  companyId?: string;
  assignedTo?: string;
  notes?: string;
  tags: string[];
  closedDate?: string;
  wonAmount?: number;
  lostReason?: string;
  tenantId: string;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
}

export interface OpportunityCreate {
  title: string;
  description?: string;
  stage?: OpportunityStage;
  amount?: number;
  probability?: number;
  expectedCloseDate?: string;
  leadId?: string;
  contactId?: string;
  companyId?: string;
  assignedTo?: string;
  notes?: string;
  tags?: string[];
}

export interface OpportunityUpdate {
  title?: string;
  description?: string;
  stage?: OpportunityStage;
  amount?: number;
  probability?: number;
  expectedCloseDate?: string;
  leadId?: string;
  contactId?: string;
  companyId?: string;
  assignedTo?: string;
  notes?: string;
  tags?: string[];
}

export interface Customer {
  id: string;
  customerId: string;
  firstName: string;
  lastName: string;
  email: string;
  phone?: string;
  mobile?: string;
  cnic?: string;
  dateOfBirth?: string;
  gender?: 'male' | 'female' | 'other';
  address?: string;
  city?: string;
  state?: string;
  country?: string;
  postalCode?: string;
  customerType: 'individual' | 'business';
  customerStatus: 'active' | 'inactive' | 'blocked';
  creditLimit: number;
  currentBalance: number;
  paymentTerms: 'Credit' | 'Card' | 'Cash' | 'Due Payments';
  assignedToId?: string;
  notes?: string;
  tags: string[];
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface CustomerCreate {
  firstName: string;
  lastName: string;
  email: string;
  phone?: string;
  mobile?: string;
  cnic?: string;
  dateOfBirth?: string;
  gender?: 'male' | 'female' | 'other';
  address?: string;
  city?: string;
  state?: string;
  country?: string;
  postalCode?: string;
  customerType?: 'individual' | 'business';
  customerStatus?: 'active' | 'inactive' | 'blocked';
  creditLimit?: number;
  currentBalance?: number;
  paymentTerms?: 'Credit' | 'Card' | 'Cash' | 'Due Payments';
  assignedToId?: string;
  notes?: string;
  tags?: string[];
}

export interface CustomerUpdate extends Partial<CustomerCreate> {}

export interface CustomerStats {
  total_customers: number;
  active_customers: number;
  inactive_customers: number;
  blocked_customers: number;
  individual_customers: number;
  business_customers: number;
  recent_customers: number;
}

export interface CustomersResponse {
  customers: Customer[];
  total: number;
}

export interface CRMLeadsResponse {
  leads: Lead[];
  pagination: {
    page: number;
    limit: number;
    total: number;
    pages: number;
  };
}

export interface CRMContactsResponse {
  contacts: Contact[];
  pagination: {
    page: number;
    limit: number;
    total: number;
    pages: number;
  };
}

export interface CRMCompaniesResponse {
  companies: Company[];
  pagination: {
    page: number;
    limit: number;
    total: number;
    pages: number;
  };
}

export interface CRMOpportunitiesResponse {
  opportunities: Opportunity[];
  pagination: {
    page: number;
    limit: number;
    total: number;
    pages: number;
  };
}

export interface CRMLeadFilters {
  status?: LeadStatus;
  source?: LeadSource;
  assignedTo?: string;
  search?: string;
}

export interface CRMContactFilters {
  type?: ContactType;
  companyId?: string;
  search?: string;
}

export interface CRMCompanyFilters {
  industry?: Industry;
  size?: CompanySize;
  search?: string;
}

export interface CRMOpportunityFilters {
  stage?: OpportunityStage;
  assignedTo?: string;
  search?: string;
}

export interface CRMMetrics {
  totalLeads: number;
  activeLeads: number;
  totalContacts: number;
  totalRevenue: number;
  projectedRevenue: number;
  conversionRate: number;
}

export interface CRMPipeline {
  stage: string;
  count: number;
  value: number;
  probability: number;
}

export interface CRMDashboard {
  metrics: CRMMetrics;
  pipeline: CRMPipeline[];
  recentActivities: any[];
  topOpportunities: Opportunity[];
  recentLeads: Lead[];
}
